### ISSUE TITLE
UI Feedback - January

### ISSUE THREAD

**Issue by MichaelPesce** [project-pareto/pareto-ui#28]
- [x] Add functionality to "zero out" a row in an input table (double click row index?)
- [x] Rework the way the app "remembers" its place when switching scenarios. When switching scenarios, the app should stay in the same section with the same filters IF those exist

**Comment by melody-shellman** [comment]
Here are notes the team has summarized. Note: the high priority items need to be completed before feb conference. The low priority is definitely not relevant until after the conference. 

- [high priority] the user units is not working as expected with case study unfortunately. Can we change the default settings to be Model Build Units:  scaled units and Scale Model: No ?
- [high priority] solver status - I see you are working with Yash/ Markus , just wanted to document it here

- [medium priority] depending on the size of the screen the user is using, the "optimize" button may not be visible, and the scroll bar may not be visible. I recall experiencing this at some point, but I can't seem to reproduce the issue. @Tominapa may be able to assist with replicating, if needed 
- [medium priority] default optimization gap can be 5% instead of 0% - Is it best if II change this on the code side?
- [medium priority] units on the input side - I can help be providing a similar dictionary as the output
- [medium priority] For the SANKEY diagram, "sourced" does not need to be plotting. There is data in the Sourced tab, but nothing plotted here

- [low priority] If the user has a case that has been optimized, if they go to optimization setup and change any value, then the status in the  scenario overview should no longer be "optimized"
- [low priority] Rename "Model Build Units" to "Model Build Units of Measure"

General Questions [low priority]
- Is it possible to "Export" my cases and have other team members upload them to their PARETO UI? The main reason would be to export a group of cases that I've created for other members of the team to use and look at . We're not looking for elegant solution here, just more curious if we can "drag and drop" data files behind the scenes

General Notes/Observations [no action needed]:
- Travis and I were creating, copying, and testing multiple scenarios (5+) at a certain point, the "optimize" button wasn't able to be clicked. We tried recopying cases that had run before, reuploading new file. What fixed it was restarting PARETO UI.
- To be aware of for team: As of now, we do not display sets in the UI. This means that a user could not use the UI to create a new completions pad, for example. 
- In the future, it could be nice for a user to create a case study entirely within the UI. No need for excel file. It would also be nice if once users complete sets, the rest of the parameters indexes are already defined.

**Comment by drouvenm** [comment]
- [ ] [medium priority] Is there any way we can "pre-load" images for selected scenarios? I've tested the "Import Network Diagram" functionality, and it works beautifully. Now I'm just wondering if we can "auto-import" images upon creating/copying scenarios. I know we talked about this but I don't remember whether we thought this was something we could do.

**Comment by drouvenm** [comment]
- [ ] [medium priority, quick fix] I find the Input Summary to be a bit confusing. Be default, we are reporting Total values for completions demand, produced water, disposal & treatment capacity. But the units for completions demand and disposal capacity are reported in [bbl/day] (which I think is incorrect considering the completions demand value). I would suggest that we remain consistent and just extrapolate daily to total capacity [bbl unit across all four metrics]. @melody-shellman: I think this is probably more something on the code side to address end than for @MichaelPesce, right?

**Comment by MichaelPesce** [comment]
@melody-shellman 
Renaming model units, removing sourced from Sankey, and changing the default optimization settings for all those mentioned (build units, scale model, optimality gap) is no problem. Only thing to note is: the default values are set when a scenario is uploaded, so changes to default settings will only be reflected in newly uploaded scenarios.

> * [ ]  - [medium priority] units on the input side - I can help be providing a similar dictionary as the output
I'm unsure what this is referring to. Do we want to display units as part of the title (ie: Completions Demand (bbl/day)?

> * [ ]  - [low priority] If the user has a case that has been optimized, if they go to optimization setup and change any value, then the status in the  scenario overview should no longer be "optimized"
This is no problem, we would still want to be able to see the previously optimized results though right? At least up until a new optimization is run

> * Is it possible to "Export" my cases and have other team members upload them to their PARETO UI? The main reason would be to export a group of cases that I've created for other members of the team to use and look at . We're not looking for elegant solution here, just more curious if we can "drag and drop" data files behind the scenes
It should be possible to copy everything you have inside <home-directory>/.pareto/* to someone else's directory, which would give them the same data that you have. It would be hard to combine two of these directories though, because IDs might overlap and the scenarios.json is a single file that stores data about all scenarios.

> * Travis and I were creating, copying, and testing multiple scenarios (5+) at a certain point, the "optimize" button wasn't able to be clicked. We tried recopying cases that had run before, reuploading new file. What fixed it was restarting PARETO UI.
The optimize button should be disabled for all scenarios when an optimization is running. Sounds like that wasn't the case here, and that this is a bug, but just wanted to clarify that. I'll have to look back over the logic I used for disabling that button

**Comment by MichaelPesce** [comment]
@drouvenm 
I can set up pre-loading images, but we will need a way for me to find the image. For example, if you upload "small_strategic_case_study.xlsx", there could be a corresponding "small_strategic_case_study.PNG" somewhere inside the Pareto codebase that would be the desired input diagram.

I'm a bit confused on the input summary issue. This could be something on my end that I misunderstood

**Comment by drouvenm** [comment]
@MichaelPesce 

> I can set up pre-loading images, but we will need a way for me to find the image. For example, if you upload "small_strategic_case_study.xlsx", there could be a corresponding "small_strategic_case_study.PNG" somewhere inside the Pareto codebase that would be the desired input diagram. 

That makes perfect sense. Let me pull together two specific examples, and then maybe we can jump on a quick call to figure out what makes the most sense. 

> I'm a bit confused on the input summary issue. This could be something on my end that I misunderstood

No worries. I'm pretty sure this is something we need to address on the code side. @melody-shellman or @tarnold17 Would you mind looking into this suggestion of mine?

**Comment by tarnold17** [comment]
> No worries. I'm pretty sure this is something we need to address on the code side. @melody-shellman or @tarnold17 Would you mind looking into this suggestion of mine?

@drouvenm I'm echoing Michael, but I also don't really understand the issue you are describing. Can you elaborate? Or perhaps  we should just discuss on a call soon. Here's what I see for the input summary, as an example:

![image](https://user-images.githubusercontent.com/26798258/215124701-affa09ef-41b7-4fd4-b8d8-8993472cbea6.png)

I think it's confusing that we see "volume" and "time" instead of actual units. Is that the issue? E.g., for "Total Completions Demand", I would expect the units to be "bbl" or "kbbl", not "volume/time".

> I can set up pre-loading images, but we will need a way for me to find the image. For example, if you upload "small_strategic_case_study.xlsx", there could be a corresponding "small_strategic_case_study.PNG" somewhere inside the Pareto codebase that would be the desired input diagram.

@MichaelPesce for the three strategic case studies in the repo - `strategic_toy_case_study.xlsx`, `strategic_small_case_study.xlsx`, and `strategic_treatment_demo.xlsx` - there are corresponding images in the repo that I added to use in the documentation. Look here: https://github.com/project-pareto/project-pareto/tree/main/docs/img

**Comment by tarnold17** [comment]
@MichaelPesce I'm playing around with the UI and I think we might have an issue with the completions demand plot. Either that, or I am interpreting it incorrectly. Here's the plot:

![image](https://user-images.githubusercontent.com/26798258/215165254-39663477-37d7-4e43-a5ab-54f9a301344c.png)

The plot shows that between T11 and T22, there is nonzero completions demand for CP01 and CP02. But if I go to the "Completions Pad Demand" tab, we can see that there actually isn't any completions demand from any of the wells for most of this stretch of time:

![image](https://user-images.githubusercontent.com/26798258/215165532-f9283ded-2bbe-4f11-b42b-96c5f88ac666.png)

**Comment by tarnold17** [comment]
@MichaelPesce The "Download an Example Input File" button downloads a file called `small_strategic_case_study.xlsx`, which has been renamed to `strategic_small_case_study.xlsx` as part of recent efforts to clean up the case studies and use a more standard naming convention for the input files: https://github.com/project-pareto/project-pareto/tree/main/pareto/case_studies

**Comment by MichaelPesce** [comment]
@tarnold17 
that's a good catch with the plot. I just tested it on my end, and there's definitely something wrong. it seems I was unintentionally disregarding 0's and empty values so the plot was missing datapoints and filling the gaps with diagonal lines. as far as I can tell, after fixing that mistake in development the plots appear to be populating correctly. I'll add that change to the latest PR

**Comment by tarnold17** [comment]
> @tarnold17 that's a good catch with the plot. I just tested it on my end, and there's definitely something wrong. it seems I was unintentionally disregarding 0's and empty values so the plot was missing datapoints and filling the gaps with diagonal lines. as far as I can tell, after fixing that mistake in development the plots appear to be populating correctly. I'll add that change to the latest PR

@MichaelPesce I'm glad we caught it. Do you know if there's a chance other plots could be similarly affected? I only noticed this issue for the completions demand plot, but I'm not sure if could affect other plots as well.

**Comment by MichaelPesce** [comment]
@tarnold17 I use the same chart component for each plot, just feed it different data. the error I fixed was inside that component, so the bug should have been affecting each plot, but the fix should now also be reflected in each plot

**Comment by melody-shellman** [comment]
>  - [medium priority] units on the input side - I can help be providing a similar dictionary as the output

@MichaelPesce, looking at units for input, it looks like I have already created a function 
`get_display_units(input_sheet_name_list, user_units)` that can help you get units for the data input. See PR #189, and particularly the `get_display_units` in `pareto/utilities/get_data.py`. You can use this function to get units to display in parentheses.

**Comment by apd-lyang** [comment]
[bug] the optimality gap option in the UI appears to be not in percentage (e.g. specifying 0.05% in the UI translates into 0.05 instead of 0.05%).

**Comment by MAZamarripa** [comment]
Gurobi solver error: Error: No available solver found among choices: (('gurobi_direct',),)
We need a stop button (after hitting run/optimize, I would like to be able to terminate the run).

**Comment by MAZamarripa** [comment]
Update text "Unfamiliar with PARETO? Check out the documentation [here](https://pareto.readthedocs.io/en/stable/)."

to: PARETO documentation can be found [here]

**Comment by tarnold17** [comment]
@MichaelPesce The UI seems to not always report the results correctly in the "Infrastructure Buildout" tab. Take the following example. In the "New Pipeline" tab, it shows that a pipeline with 4 inch diameter has been built from PP05 to N12:

![image](https://user-images.githubusercontent.com/26798258/215897292-c0f9cfa3-300e-4367-b414-d02c53e0100f.png)

However, this entry is absent from the "Infrastructure Buildout" tab:

![image](https://user-images.githubusercontent.com/26798258/215897420-4c519463-e991-4af0-8fc8-119eaeb222ce.png)

**Comment by MichaelPesce** [comment]
@tarnold17 I don't build that table, just display it
@melody-shellman do you know what's going on here

**Comment by melody-shellman** [comment]
@tarnold17, this CAPEX chart is built on our end. Can you help point me to a  case and  config settings that replicate this issue? There is a check in the code that only adds the CAPEX to the overview table if the value of the binary is equal to 1. My guess is that, sometimes, it's equal to 0.9999 because of scaling, and it isn't included. 
I'd like to test this out with an example, if you have one.

**Comment by apd-lyang** [comment]
When switching from one scenario to another, if the first scenario has a table that's empty, then the table in the second scenario is empty immediately after the switch, even if the table is not actually empty.  User needs to click on another table first and switch back to the table of interest for the result to show properly. 

Example: 
Pipeline shutdown scenario (has table for "Trucked")
![image](https://user-images.githubusercontent.com/104799444/216130102-f60f876b-4112-4496-a4c4-7ad9f69caae4.png)
Switch to base scenario where "Trucked" table is empty
![image](https://user-images.githubusercontent.com/104799444/216130144-660ab875-8fd9-4353-88da-1b580188bb43.png)
Switching back to shutdown scenario where "Trucked" now appears to be empty
![image](https://user-images.githubusercontent.com/104799444/216130382-c01e5ec4-367d-414b-a450-eb9755a97cd7.png)
Click on another table and click back to "Trucked", now table is shown properly again
![image](https://user-images.githubusercontent.com/104799444/216130460-4e83addd-4540-4e4c-9fe9-b0b26fad9fa6.png)

**Comment by apd-lyang** [comment]
Multi-column filtering doesn't work as intended, the row filter only shows the first field (repeated if multi-indexed) for all rows and does not show the selected column (besides the first column) that needs to be filtered.
![image](https://user-images.githubusercontent.com/104799444/216131347-f30b143d-153b-4121-b617-810a909f9394.png)
